Support params/buffers inside cond and map

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103861
* __->__ #102310

With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim

tugsbayasgalan committed:
Support params/buffers inside cond and map

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+                    ret_val, ret_graph, ret_lifted_freevars = speculate_subgraph(
+                        tx, args[ix], operands, graph_checkpoint, checkpoint
+                    )
+                except Exception as e:
+                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
```
This codepath gets used by both training and export. Just to check, is UserError an export-specific thing, or does it work with training as well?

The thing I am concerned about is that speculate_subgraph can `raise Unsupported`, which would lead to a graph break in training. However, that exception would be caught by this general `Exception`.  Does this end up squashing the graph break in training?

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
-            ) = speculate_branch(False)
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+
+            tx.output.side_effects.prune_dead_object_new(tx)
```
I'm not sure this is what we want. In my experience the side effect pruner will end up pruning too much because it is meant to be called when Dynamo ends up compiling the subgraph (and so it ignores things like free variables).

tugsbayasgalan replied:
Yeah i was just trying to preserve the old behaviour. Can just delete i guess.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+
+            tx.output.side_effects.prune_dead_object_new(tx)
+            state_after_true_branch = tx.copy_graphstate()
```
We should be asserting that before and after speculate_subgraph, the side effects are the same

tugsbayasgalan replied:
Added a check

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1092,16 +1007,39 @@ def speculate_branch(branch):
+                            break
+                        inp_count += 1
+
+            fixup_branch_inps(
+                true_graph,
+                len(operands) + len(true_lifted_freevars) - 1,
+                false_lifted_freevars,
+                "_false_branch",
+            )
+
+            fixup_branch_inps(
+                false_graph, len(operands) - 1, true_lifted_freevars, "_true_branch"
+            )
```
What happens if both true and false branches capture the same variable? Should add a test for this.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1092,16 +1007,39 @@ def speculate_branch(branch):
             )
 
-            # Apply side effects (guaranteed to be equal)
+            # TODO do we need this?
```
We shouldn't, speculate_subgraph writes new side effects directly into tx.output.side_effects

tugsbayasgalan replied:
ok will delete

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            state_after_true_branch = tx.copy_graphstate()
+            true_guards = state_after_true_branch.output.guards
+            true_nn_modules = state_after_true_branch.output.nn_modules
+            true_cmp = get_comparable_state(state_after_true_branch)
```
Is the `true_cmp` object actually used for anything? Ditto for `false_cmp`.

tugsbayasgalan replied:
Nah just trying to preserve the old code.

tugsbayasgalan replied:
They are used to add tracking tensors to global tensor tracking list. So it seems needed?

zou3519 replied:
Where does that logic happen?

tugsbayasgalan replied:
https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/torch.py#L1138

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1130,22 +1070,30 @@ def speculate_branch(branch):
                 checkpoint,
             )
 
+            tx.output.side_effects.prune_dead_object_new(tx)
+            state_after_body = tx.copy_graphstate()
+            body_guards = state_after_body.output.guards
+            body_nn_modules = state_after_body.output.nn_modules
+            body_cmp = get_comparable_state(state_after_body)
```
Same questions from the cond implementation apply here

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -3124,6 +3110,136 @@ def forward(self, x):
+        gm, _ = torch._dynamo.export(M(), torch.ones(6, 4), aten_graph=False)
+        self.assertEqual(gm(torch.ones(6, 4)), M()(torch.ones(6, 4)))
+        self.assertEqual(gm(torch.ones(3, 4)), M()(torch.ones(3, 4)))
+
```
All of the tests are for export. We should add tests for training as well under test/dynamo/test_higher_order_ops.py.

I'm not really sure what the best way to prevent overlap these test_export and test_higher_order_ops is. Morally tests for training should be a superset of the tests for export, so I would actually toss all of the heavy testing into test_higher_order_ops, but it sounds like export goes through a slightly different mechanism?

Things we should test:
- map (we can go light on these tests, because we have a straightforward application of speculate_subgraph)
   - map with captured value
   - map with side effects in body
   - map with graph break in body
- cond (there are a bunch of interesting tests here)
    - graph break in true or false branch
    - true branch has free variables
    - false branch has free variables
    - true and false branch both have free variables
    - true and false branch capture free variables (with some overlap)
    - true branch has side effects
    - false branch has side effects
    - only the true branch or the false branch has a guard (I'm not really sure how to write this test, but we want to make sure that this actually invokes a recompile or turns into a graph break if it's not implemented)

tugsbayasgalan replied:
Yep I think I covered majority of the cases.

zou3519 reviewed (commented):
Some high-level comments:
- Handling of side effects looks a bit suspicious
- I don't see where we are comparing true_cmp/false_cmp, so these don't seem necessary? (though I probably just misread something)
- we should add more tests, especially on the torch.compile side

tugsbayasgalan commented on torch/_dynamo/variables/torch.py:
```diff
@@ -781,7 +780,8 @@ def speculate_subgraph(
                 if not isinstance(
-                    output, (TensorVariable, ListVariable, TupleVariable)
+                    output,
+                    (TensorVariable, ConstantVariable, ListVariable, TupleVariable),
```
Good point! I saw a test case that didn't respect this contract. (https://github.com/pytorch/pytorch/blob/68e55bff62fd0f0458bd252dd5a09315530390c1/test/dynamo/test_export.py#L3001) Will rewrite it to Tensor.

tugsbayasgalan replied:
Actually this test is checking that we can not return const. Reverted this change

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

avikchaudhuri commented:
LGTM. @tugsbayasgalan have you addressed all of @zou3519's comments? If so can you re-request a review?

avikchaudhuri reviewed (approved):
Pending approval from @zou3519

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46344826).

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1130,22 +1070,30 @@ def speculate_branch(branch):
                 args[0],
                 [
-                    get_fake_value(args[1].as_proxy().node, tx)[0],
+                    first_dim,
```
Ohhh does this do what happens on line 1125?

ydwu4 replied:
Yes!

zou3519 commented on functorch/experimental/_cond.py:
```diff
@@ -125,11 +125,34 @@ def cond_dense(pred, true_fn, false_fn, operands):
+        # If there is requires_grad, we delay the error until backward pass
+        if requires_grad:
+            # cond can only return one value
+            err_fn = torch._C._functions.DelayedError(
+                b"trying to differentiate buffers inside torch.cond",
+                1,
+            )
```
nit: I'd phrase this as "NYI: torch.cond doesn't support autograd"

tugsbayasgalan replied:
will do

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -282,13 +282,183 @@ def false_fn2(x, y):
+                    return x.sum() + self.buffer.sum() + z.sum()
+
+                def false_fn(x):
+                    return x.sum() - self.buffer.sum() - z.sum()
```
To change things up, can we swap the order that `self.buffer` and `z` appears in false_fn?

tugsbayasgalan replied:
Yep

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -282,13 +282,183 @@ def false_fn2(x, y):
+        mod_for_compile = torch.compile(Foo(), backend=cnt, dynamic=True)
+        mod_for_eager = Foo()
+
+        self.assertEqual(
+            mod_for_compile(torch.ones(6, 4)), mod_for_eager(torch.ones(6, 4))
+        )
+
```
We should also assert that dynamo did graph break on this, by using `cnt` to count the number of frames. I'm not really sure what the number is, but the important thing is that it is > 1.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1096,32 +1005,40 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
-            (
-                true_r,
-                true_graph,
-                true_guards,
-                true_nn_modules_context,
-                true_cmp,
-            ) = speculate_branch(True)
-            (
-                false_r,
-                false_graph,
-                false_guards,
-                false_nn_modules_context,
-                false_cmp,
-            ) = speculate_branch(False)
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+
+            state_after_true_branch = tx.copy_graphstate()
+            true_guards = state_after_true_branch.output.guards
+            true_nn_modules = state_after_true_branch.output.nn_modules
+            true_cmp = get_comparable_state(state_after_true_branch)
+
+            (false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
+
+            state_after_false_branch = tx.copy_graphstate()
+            false_guards = state_after_false_branch.output.guards
+            false_nn_modules = state_after_false_branch.output.nn_modules
+            false_cmp = get_comparable_state(state_after_false_branch)
```
A lot of this logic is redundant and we should just delete most of it. `speculate_subgraph` already (by design) mutates tx.output's state (guards, tracked_fakes, tracking), so there should be no need for us to add additioanl things to it.

This function should just look like:
```
(true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
(false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
fixup_branch_inps(true_branch)
fixup_branch_inps(false_branch)
add_subgraph(true_branch)
add_subgraph(false_branch)
```
If you agree, then this also should apply to the map() implementation below.

Also, I'm not really sure how to test that we've propagated the guards/fake_tensor values/etc correctly. We should probably add a test that we're not regressing that

tugsbayasgalan replied:
Yep seems to be working after I delete all these.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1142,26 +1059,51 @@ def speculate_branch(branch):
+            # FIXME (tmanlaibaatar) this is probably wrong?
+            tx.output.side_effects = true_cmp.output.side_effects
+
+            def fixup_branch_inps(graph, add_after, new_args, suffix) -> None:
+                inp_count = 0
+                for node in graph.nodes:
+                    if node.op == "placeholder":
+                        if inp_count == add_after:
+                            with graph.inserting_after(node):
+                                for inp_node in new_args:
+                                    new_node_name = inp_node.node.name + suffix
+                                    graph.placeholder(new_node_name)
+                            break
+                        inp_count += 1
```
Can you add a comment on what this is necessary for? IIRC this is to handle a case where if a free variable only exists in one of the branches, then we need to promote it to being inputs to both branches.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1142,26 +1059,51 @@ def speculate_branch(branch):
+                            break
+                        inp_count += 1
+
+            fixup_branch_inps(
+                true_graph,
+                len(operands) + len(true_lifted_freevars) - 1,
+                false_lifted_freevars,
+                "_false_branch",
+            )
+
+            fixup_branch_inps(
+                false_graph, len(operands) - 1, true_lifted_freevars, "_true_branch"
+            )
```
I'm not convinced that this code is sufficient to handle overlapping free variables in both branches. (I added suggestions for more tests in another comment)

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -282,13 +282,183 @@ def false_fn2(x, y):
+                            if node.op == "placeholder"
+                        ]
+                    )
+                    self.assertEqual(num_placeholders, 5)
```
I'm a bit nervous about free variables in different branches of cond, so we should add more test cases around those. What I want to see are things like:

- true_fn(x) captures (a, c, d), false_fn(x) captures (a, b, d). We want to make sure that the free variable lifting logic rewrites both true_fn and false_fn to accept (a, c, d, b) (or something reasonable). Furthermore, we should use self.assertExpectedInline so that we can actually have what the graphs look like in the code so we know what we're working with.
- true_fn(x) captures (a, c, b), false_fn(x) captures (b, a, c). The result is that Dynamo should rewrite both branches to accept (x, a, c, b) (or something reasonable). Furthermore, we should use self.assertExpectedInline so that we can actually have what the graphs look like in the code so we know what we're working with.

This is a bit of a mini-algorithms problem, given two lists of free variables, how do we unify them so that both functions accept the list of all of the variables in the same order.

Bonus points if we can write these tests in a more compact form so that we don't need to write 50 LOC to test each of these cases, because there are probably more edge cases to test than the ones I've listed above

zou3519 reviewed (commented):
This looks mostly good to me, but I had some high-level points:
- I think a lot of the code now doesn't do anything; we should clean it up so that it is clearer what is going
- We should test what is going on with the cond free variables situation better. In particular, if the two branches both capture a set of free variables, and there is some overlap between them, what happens?

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -242,14 +242,24 @@ def dispatch(self, dispatch_key, *args, **kwargs):
-        if torch.overrides.has_torch_function(flat_args):
-            return torch.overrides.handle_torch_function(
-                self, flat_args, *args, **kwargs
+        # Dynamo already traces the body of HigherOrderOp beforehand when it
+        # so no need to trace into it.
+        from torch._dynamo.eval_frame import disable
```
This assumes that `torch._dynamo` is already imported, which it might not be. We need to add an `import torch._dynamo` before this line

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1096,72 +1005,78 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            (false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
+            false_nn_modules = tx.copy_graphstate().output.nn_modules
+
+            # TODO (tmanlaibaatar) deduplicate this later
```
Could you file an issue about this so we can track it? Ideally we would link to the issue in this comment.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1180,45 +1095,37 @@ def speculate_branch(branch):
             # To get the example output from map() we will need to prodive at least one sample to
             # the loop body. In our case we will always use xs[0], and our map() won't support zero
             # sized tensor during tracing.
+            first_dim = args[1].call_method(
```
nit: can you change `prodive` to `provide` in the above comment haha

zou3519 commented on functorch/experimental/_cond.py:
```diff
@@ -125,11 +125,34 @@ def cond_dense(pred, true_fn, false_fn, operands):
+        result = cond(pred, true_fn, false_fn, *operands)
+
+        # If there is requires_grad, we delay the error until backward pass
+        if requires_grad:
+            # cond can only return one value
+            err_fn = torch._C._functions.DelayedError(
+                b"NYI: torch.cond doesn't support autograd",
+                1,
+            )
```
Could you add a test that this gets triggered on backwards?

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -3113,6 +3047,205 @@ def forward(self, x):
+
+        out_graph, _ = torch._dynamo.export(mod, pred_x, x)
+        self.assertExpectedInline(
+            out_graph.code.strip(),
```
nit: Is it possible to assertExpectedInline on out_graph.print_readable() ? (without it actually printing out?) This way we can also look at what the cond true_fn/false_fn graph looks like

tugsbayasgalan replied:
Yep the bottom asserts are already testing for branch graphs

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -282,13 +282,203 @@ def false_fn2(x, y):
+                    )
+                    self.assertEqual(num_placeholders, 5)
+
+    def test_cond_side_effect_in_one_branches(self):
```
Maybe mark this with an `expectedFailure` and instead `self.assertEqual(len(backend.graphs), 0)`? Or add a comment about this not supposed to be supported but is at the moment.

zou3519 reviewed (approved):
LGTM, thanks for improving the testing and cleaning up the code! I had some minor last-mile comments

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D46344826](https://our.internmc.facebook.com/intern/diff/D46344826)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -ic

tugsbayasgalan commented:
@pytorchbot merge -i

tugsbayasgalan commented:
/easycla

tugsbayasgalan commented:
@pytorchbot merge -i

tugsbayasgalan commented:
@pytorchbot merge -i

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46746202).

huydhn commented:
@pytorchbot revert -m 'The test is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/766f236bad2327060575780219e0d4964dc661e5' -c landrace

avikchaudhuri commented:
Ouch, one failing test...

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

Differential Revision: [D46746202](https://our.internmc.facebook.com/intern/diff/D46746202)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

Differential Revision: [D46746202](https://our.internmc.facebook.com/intern/diff/D46746202)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

Differential Revision: [D46746202](https://our.internmc.facebook.com/intern/diff/D46746202)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "Support params/buffers inside cond and map"


With #102022, params and buffers are always treated as special case of free variables. In this PR, I switch cond and map implementation to the this method and deprecate the old tracing mechanism. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

Differential Revision: [D46746202](https://our.internmc.facebook.com/intern/diff/D46746202)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge
